interfaces: #30697 follow ups

ismaelsadeeq opened the pull request:
This PR addresses the remaining review comments from #30697

1. Disallowed overwriting settings values with a `null` value.
2. Uniformly used the `SettingsAction` enum in all settings methods instead of a boolean parameter.
3. Updated `overwriteRwSetting` to receive the `common::SettingsValue` parameter by value, enabling it to be moved safely.
4. Removed wallet context from the `write_wallet_settings_concurrently` unit test, as it is not needed.

ismaelsadeeq committed:
test: remove wallet context from `write_wallet_settings_concurrently`

ismaelsadeeq committed:
chain: move new settings safely in `overwriteRwSetting`

ismaelsadeeq committed:
chain: uniformly use `SettingsAction` enum in settings methods

sedited reviewed (commented):
Concept ACK

TheCharlatan commented on src/interfaces/chain.h:
```diff
@@ -361,7 +361,7 @@ class Chain
 
     //! Replace a setting in <datadir>/settings.json with a new value.
-    virtual bool overwriteRwSetting(const std::string& name, common::SettingsValue& value, bool write = true) = 0;
+    virtual bool overwriteRwSetting(const std::string& name, common::SettingsValue value, bool write = true) = 0;
```
Ah, I see. Please resolve :)

ismaelsadeeq replied:
Thanks

fanquake commented:
cc @stickies-v @ryanofsky

ryanofsky commented on src/node/interfaces.cpp:
```diff
@@ -819,18 +819,22 @@ class ChainImpl : public Chain
     bool overwriteRwSetting(const std::string& name, common::SettingsValue value, interfaces::SettingsAction action) override
     {
-        if (value.isNull()) return deleteRwSettings(name, action);
+        if (value.isNull()) return false; // disallow overwriting with null values.
```
In commit "chain: disallow overwriting settings with null values" (4627174d80838e1c6848f9dff47a2ca4a9a35091)

Would be good to just delete this line entirely and make overwriteRwSetting just call updateRwSetting without making a special case for null. Setting a null setting should just delete a setting consistently and not have different behavior for different functions.

(Sorry I didn't notice this line before and mistakenly thought overwriteRwSetting would allow writing null values to the settings file, so my earlier comments might have been confusing.)

furszy replied:
> Setting a null setting should just delete a setting consistently and not have different behavior for different functions.

I would rather fail when a null setting is provided to any of the update or overwrite functions. I believe the intent to delete should be explicit, at least within this settings layer. That's why the `deleteRwSettings` function was introduced. It seems inconsistent to have two different methods for deleting a value - one embedded within the update function and another explicit.

ryanofsky replied:
So we both agree it is bad for overwrite function to reject null values and update function to accept them, but I would prefer both functions to accept them and you want both functions to reject them.

Here are the reason I think it is better to accept them:

- I think it is generally bad practice to accept nullable values and then handle them with runtime errors. If you want to write API's that don't accept null values I think you should use non-nullable types.

- I think in this specific case, there is no reason these APIs should not accept null values. Ihe purpose of the update function is to be able to atomically read and update settings. But there is no function to atomically read and delete settings and this should be a completely valid use-case for situations where you want to safely reset settings without clobbering concurrently set values. We can easily support it by just not going out of the way to add unnecessary restrictions.

I am fine with existence of deleteRwSetting as a more convenient & readable wrapper around overwriteRwSetting and updateRwSetting, the same way overwriteRwSetting is a convenient and readable wrapper around updateRwSetting. But existence of deleteRwSetting should not be an excuse for crippling the behavior of the other functions and making them less safe to call. Ability to atomically read and delete settings is useful, and trying to deal with null values at runtime rather than compile time is a bad practice that should be a last not first resort.

ismaelsadeeq replied:
There are two suggestions here:

1. **@furszy suggested to fail when the value is null**: I agree with not checking for null values, for the reasons outlined above.
   
2. **@ryanofsky suggested removing the null value check**: This approach would overwrite the previous settings with a null value.
   
> Setting a null value should consistently delete the setting, rather than having different behavior in different functions.

However, if we don’t delete the key-value pair, the setting still exists in the map with a null value, which does not truly remove the setting, right?

Currently, the code calls `deleteRwSettings` when the `value` is null, which deletes the setting entirely. This makes the method redundant, as calling `deleteRwSettings` achieves the same result as passing a null `SettingsValue` to `overwriteRwSetting`.

I believe a better approach is to simply update the settings based on what is passed to `overwriteRwSetting`.
It should be the caller's responsibility to determine whether to pass a null value or not. 
If the caller intends to delete the setting, they should explicitly call `deleteRwSettings`.

ryanofsky replied:
> I think the only thing that "bothers" me a bit about the nullable class is the different meanings it could have, which require additional clarifications at the API level. 

Yeah this is real concern, and I think a common thing that happens with null values, true/false values, empty strings, and special integer values. If you are sloppy with these values you will run into problems. If you are thoughtful you can come up with very clean interfaces, but it is hard to be careful. You can avoid problems with ambiguous values by using enums, std::variant, not_null implementations, and custom types, but these can be inconvenient or verbose. We're just trying to pick a tradeoff that is convenient and doesn't cause big surprises.

ismaelsadeeq committed:
chain: dont check for null settings value in `overwriteRwSetting`

- Just call updateRwSetting it will erase the settings when the new
  value is null.

ismaelsadeeq commented on src/node/interfaces.cpp:
```diff
@@ -828,20 +828,20 @@ class ChainImpl : public Chain
             settings.rw_settings.erase(name);
         });
-        return !write || args().WriteSettingsFile();
+        return action == interfaces::SettingsAction::SKIP_WRITE || args().WriteSettingsFile();
```
I guess you are trying to avoid a situation where we would need to write:

```
return action == interfaces::SettingsAction::SKIP_WRITE || action == interfaces::SettingsAction::NEW_ACTION || args().WriteSettingsFile();
```

In some cases we would still need to modify this to handle the new action if it has additional effects beyond just writing, such as writing and logging, for example.

---
I've updated to your suggestion, as it is more maintainable.

ismaelsadeeq committed:
chain: ensure `updateRwSetting` doesn't update to a null settings

Co-authored-by: Ryan Ofsky <ryan@ofsky.org>

ryanofsky reviewed (commented):
Code review ACK adc019860d7df6848b98aec58f1c3ff47936e06b. These are nice, well-organized improvements to the settings interface to make it more consistent, remove ability to write null values to settings.json, and add ability to atomically read and delete settings.

I think not allowing null values in `settings.json` is good change to encourage clearer representations of settings, maintain parity between settings.json / bitcoin.conf / command line settings representations, avoid making ArgsManager implementation more complicated, and helpful if we want to have a clean RPC interface for updating settings later, since our JSON-RPC framework does not distinguish between null and unspecified parameter values, and bitcoin-cli doesn't provide a clean way to pass null parameters.

The PR looks good in it's current form, but could consider improving documentation and simplifying the `deleteRwSettings` implementation with:

```diff
--- a/src/interfaces/chain.h
+++ b/src/interfaces/chain.h
@@ -356,14 +356,21 @@ public:
     virtual common::SettingsValue getRwSetting(const std::string& name) = 0;

     //! Updates a setting in <datadir>/settings.json.
+    //! Null can be passed to erase the setting. There is intentionally no
+    //! support for writing null values to settings.json.
     //! Depending on the action returned by the update function, this will either
     //! update the setting in memory or write the updated settings to disk.
     virtual bool updateRwSetting(const std::string& name, const SettingsUpdate& update_function) = 0;

     //! Replace a setting in <datadir>/settings.json with a new value.
+    //! Null can be passed to erase the setting.
+    //! This method provides a simpler alternative to updateRwSetting when
+    //! atomically reading and updating the setting is not required.
     virtual bool overwriteRwSetting(const std::string& name, common::SettingsValue value, SettingsAction action = SettingsAction::WRITE) = 0;

     //! Delete a given setting in <datadir>/settings.json.
+    //! This method provides a simpler alternative to overwriteRwSetting when
+    //! erasing a setting, for ease of use and readability.
     virtual bool deleteRwSettings(const std::string& name, SettingsAction action = SettingsAction::WRITE) = 0;

     //! Synchronously send transactionAddedToMempool notifications about all
--- a/src/node/interfaces.cpp
+++ b/src/node/interfaces.cpp
@@ -841,10 +841,7 @@ public:
     }
     bool deleteRwSettings(const std::string& name, interfaces::SettingsAction action) override
     {
-        args().LockSettings([&](common::Settings& settings) {
-            settings.rw_settings.erase(name);
-        });
-        return action != interfaces::SettingsAction::WRITE || args().WriteSettingsFile();
+        return overwriteRwSetting(name, {}, action);
     }
     void requestMempoolTransactions(Notifications& notifications) override
     {
```

Other followups are also possible and could be future PRs:

1. Settings methods in `interface::Chain` and `interfaces::Node` should probably be dropped and moved into an `interfaces::Settings` class returned by an `interfaces::init::makeSettings()` method.
2. It could make sense to move some or all of the `updateRwSetting` implementation into `src/common/args` or `src/common/settings` as an ArgsManager method or helper function.
3. It could make sense to add unit test coverage in `src/test/interfaces_tests.cpp`.

ismaelsadeeq committed:
chain: simplify `deleteRwSettings` code and improve it's doc

Co-authored-by: Ryan Ofsky <ryan@ofsky.org>

ismaelsadeeq commented:
> The PR looks good in it's current form, but could consider improving documentation and simplifying the deleteRwSettings implementation with

Done in the latest force push see diff https://github.com/bitcoin/bitcoin/compare/adc019860d7df6848b98aec58f1c3ff47936e06b..84663291275248fd52da644b0c2566bbf9cc780b

ryanofsky commented on src/node/interfaces.cpp:
In commit "chain: dont check for null settings value in `overwriteRwSetting`" (f8d91f49c7091102138fcb123850399e8fadfbc7)

Might be good to say in commit message that this change is a refactoring and does not change behavior. Could have "refactor:" prefix in the title.

ismaelsadeeq replied:
thanks will do if there is need to retouch.

ryanofsky commented on src/node/interfaces.cpp:
```diff
@@ -841,10 +841,7 @@ class ChainImpl : public Chain
-            settings.rw_settings.erase(name);
-        });
-        return action != interfaces::SettingsAction::WRITE || args().WriteSettingsFile();
+        return overwriteRwSetting(name, {}, action);
```
In commit "chain: simplify `deleteRwSettings` code and improve it's doc" (84663291275248fd52da644b0c2566bbf9cc780b)

Might be good to say in commit message that this change is a refactoring and does not change behavior. Could have "refactor:" prefix in the title.

ismaelsadeeq replied:
I will do it when their is need to retouch.

ryanofsky reviewed (approved):
Code review ACK 84663291275248fd52da644b0c2566bbf9cc780b. Looks good, thanks for taking suggestions and applying them to the right commits. Only changes since last review were documentation improvements and simplifying delete method.

furszy reviewed (commented):
Code review ACK 84663291275248fd52da644b0c2566bbf9cc780b

It seems that with the latest changes, the `deleteRwSettings` is no longer used.
It would be nice to add test coverage for it and for the null value usage as well.
